lib/Bailador/Request.pm

```diff
@@ -25,40 +27,61 @@ class Bailador::Request {
     }
     multi method params ($source) {
         my %ret;
+        my $source_data = '';
         given $source {
             when 'query' {
-                if $!env<QUERY_STRING> {
-                    for $.env<QUERY_STRING>.split('&') -> $p {
-                        my $pair = $p.split('=', 2);
-                        %ret{$pair[0]} = uri_unescape $pair[1];
-                    }
-                }
+                $source_data = $.env<QUERY_STRING> if $!env<QUERY_STRING>;
             }
             when 'body' {
-                if $!env<p6sgi.input> {
-                    for $.env<p6sgi.input>.decode.split('&') -> $p {
-                        my $pair = $p.split('=', 2);
-                        %ret{$pair[0]} = uri_unescape $pair[1];
-                    }
-                }
+                $source_data = $.env<p6sgi.input>.decode if $!env<p6sgi.input>;
             }
             default {
                 die "unknown source '$source'";
             }
         }
+        return %ret unless $source_data.chars;
+
+        for $source_data.split('&') -> $p {
+            my @pair = $p.split('=', 2);
+            if (%ret{@pair[0]}:exists) {
+                %ret.push((@pair[0] => |uri_unescape(@pair[1])));
+            }
+            else {
+                %ret{@pair[0]} = uri_unescape @pair[1];
+            }
+        }
 
         return %ret;
     }
 
+    method headers () {
+        return %!headers if %!headers;
+        for $.env.keys.grep(rx:i/^[HTTP||CONTENT]/) -> $key {
+            my $field = $key;
+            $field ~~ s:i/^HTTPS?_//;
+            %!headers{$field} = $.env{$key};
+        }
+        return %!headers;
+    }
+
     method cookies () {
-        my %ret;
-        if $.env<HTTP_COOKIE> {
-            for $.env<HTTP_COOKIE>.split('; ') {
-                my ($name, $value) = $_.split('=', 2);
-                %ret{uri_unescape($name)} = uri_unescape($value);
+        return %!cookies if %!cookies || !$.env<HTTP_COOKIE>;
+        for $.env<HTTP_COOKIE>.split(/<[;,]>\s/) -> $cookie {
+            my ($name, $value) = $cookie.trim.split(/\s*\=\s*/, 2);
+            my @values;
+            if $value {
+                @values = $value.split(/<[&;]>/).map: { uri_unescape($_) };
             }
+            # TODO: build Bailador::Cookie object :)
+            %!cookies{uri_unescape($name)} = @values;
         }
-        return %ret;
+        return %!cookies;
+    }
+
+    method is_ajax returns Bool {
+        return True if $.header<X-Requested-With>:exists
+                    && $.header<X-Requested-With> eq 'XMLHttpRequest';
```

Comment:
You're right! And not just here: RFC 2616 says header field names should be case insensitive. I'll make a patch and amend the PR, thanks!